### ISSUE TITLE
Improved view support

### ISSUE THREAD

**Issue by ilmax** [aspnet/EntityFrameworkCore#827]
Hi, I have prototyped a really basic (or even less) view support through part of the EF7 stack, currently my prototype is able to
- Map entity to a View via fluent api
- Create a view via migrations
- Read and materialize values backed by views
- Avoid updating entities if backed by views

As stated above, this is a really naive approach and I know there are a tons of unsupported features at the moment, anyway I really would like to improve this further and I would be glad to discuss implementation details with the EF team and the community.

I would like to know if the EF team are interested in this contributions.
Thanks, Max.

**Comment by ilmax** [comment]
Hi @rowanmiller, does the team is interested in such contribution? I would like to be able to query views with EF, and would love to give my contribution on this feature.

If this is something the team can be interested in, I would like to start a discussion around the needed changes and the general direction to take.

Regards, Max.

**Comment by milutinovici** [comment]
I agree with Max, this would be an excellent feature. It would be even cooler if we could use linq to define indexed views.

**Comment by rogerhendriks** [comment]
Without view support EF7 is almost useless as, I'm sorry to say, the query generator does not have enough intelligence for supporting high performant business applications with 10 or more joins, unions and complex "group by" statements on applications with millions of records. And some databases just all ready have good views on board. ICT work is often about modernizing gui's on old databases. 
And we don't want to use raw sql to be able to get a 1-to-1 mapping of views/viewmodels. Please do support views and updatable views by -self defining- keys in metadata as we now have to use 6.x isnull/nullif and 'editing the edmx after each update' workarounds.

**Comment by staff0rd** [comment]
@ilmax care to share how you are mapping views and generating migrations for them?

**Comment by ilmax** [comment]
@staff0rd my prototype was build against early ef7 version, since ef was in the early days at the time, I abandoned the prototype, waiting for ef7 to  RC.
I think that now the majority of EF building blocks are quite stable, so maybe I can start working on this and put together a prototype here on github.

**Comment by gdoron** [comment]
@ilmax any update, EF 7 is RC already... :)
I would really be happy using this feature.

**Comment by ilmax** [comment]
@gdoron Since EF Core is a completely new rewrite, over the time a lot of parts have changed so my POC is not working anymore, I plan to work on this as soon as EF Core reaches RTM

**Comment by gdoron** [comment]
@ilmax I think since EF is RC already there won't be as much changes as when it was Alpha/beta.

But regardless, once you contributed the code and it was merged, it's in the repository, and if the EF team will do a breaking change that break the view generation (not a likely scenario since RC), they will have to fix it, so I wouldn't be too much worry about it.

**Comment by rowanmiller** [comment]
BTW in the meantime you can just tell EF that the view is a table. The limitations of this approach are that you need to give EF some properties to use as a key, and migrations won't create a view for you. You can also use raw SQL to select from a view `context.Blogs.SqlQuery("SELECT * FROM dbo.TopBlogsView")`.

**Comment by rowanmiller** [comment]
> But regardless, once you contributed the code and it was merged, it's in the repository, and if the EF team will do a breaking change that break the view generation (not a likely scenario since RC), they will have to fix it, so I wouldn't be too much worry about it.

This is true, but the feature needs to be completely implemented (including tests etc.) for us to take the code.

**Comment by gdoron** [comment]
@rowanmiller but how can I trick EF not to try and create the table (which is a view) in the migrations?
Manually remove the creation from the migration? 

> context.Blogs.SqlQuery("SELECT * FROM dbo.TopBlogsView").

This will work only if the view returns data from a single table, if it's (like in 99% of the times) a view that returns data from multiple composed tables, we'll get only the data from the Blogs, the columns from _Users_ let's say will be ignored when deserialized. 

> in the meantime you can just tell EF that the view is a table

You reminded me an [old jewish joke](http://www.jokes4us.com/religiousjokes/everyfridayjoke.html) I heard 10 years ago. :)

**Comment by rowanmiller** [comment]
>  but how can I trick EF not to try and create the table (which is a view) in the migrations?
 Manually remove the creation from the migration? 

Yes, that is what you would need to do

**Comment by gdoron** [comment]
@rowanmiller I don't see VIEWs support in the roadmap at all.
Do you really consider it be in in such a low priority that all the stuff in the roadmap are more important than VIEWs or you simply forgot to mention it?

Thanks

**Comment by rowanmiller** [comment]
We just missed it off the list, I've added it to the roadmap.

**Comment by gdoron** [comment]
@rowanmiller You might want to add a link in the roadmap to this PR for how it can be done in the meanwhile:  #4561 so people won't think there's absolutely no way of doing it.

Thanks

**Comment by rogerhendriks** [comment]
@rowanmiller great to see it's on the roadmap! Kutgw

**Comment by vcannaday** [comment]
Has this issue been targeted for a release yet? 

We are currently having to maintain two database catalogs for code generation. We have one database that contains table schema only to represent the views that exist in our actual database.

Vance

**Comment by douglasg14b** [comment]
Any news on this? I'm having a hard time justifying continued EF Core use without view capabilities...

**Comment by AxelAndersen** [comment]
Do you have any release date for supporting views?

**Comment by shaulbehr** [comment]
For the record, I'll add a "Me, too!" comment here.  Please can you prioritize View support?

**Comment by rowanmiller** [comment]
One of the big things folks want when working with views is to be able to query data into objects that are not entities (and therefore the view doesn't have to be normalized or have a key). We expect to at least make progress on that feature in 2.0 (https://github.com/aspnet/EntityFramework/issues/1862).

**Comment by BillieBishop** [comment]
When can we hope to see it done?

**Comment by ErikEJ** [comment]
@BillieBishop Did you see https://github.com/aspnet/EntityFramework/issues/4561 ?

**Comment by danobri** [comment]
I agree with many others here - view support is critically important - please try to get it in the next release.  The lack of view support is especially painful given that there are also issues with doing projections on related objects - #5738 and #7131.  Unless there is something I am missing - at this point the only efficient way to do a query that joins a bunch of entities and selects a minimal number of fields is to pass raw SQL (i.e. - SqlQuery function).

**Comment by gsGabriel** [comment]
@danobri you can map the views...
You can do this manually, you will need to create a class that reflects your view and map it in your context. It Works very well for me.

**Comment by ArgusMagnus** [comment]
> For the record, I'll add a "Me, too!" comment here. Please can you prioritize View support?
Me too

**Comment by nhustak** [comment]
I'll add to the  'I require views' and 'it's about useless to me without it'

**Comment by rnanjala** [comment]
i have installed entiry core 2.0 version and it's still not supporting views. any timeline as when we can use views using entitycore?

**Comment by ajcvickers** [comment]
@nhustak @rnanjala @ArgusMagnus Can you be more specific about what you want with regards to "supporting views"--perhaps looking at #9290 as well. The reason I ask is that there are various ways that EF could interact with views, but taking two examples, the work to reverse engineer from views is very different from the work to query for data without keys, and both of these could be considered "view support".

**Comment by nhustak** [comment]
@ajcvickers  I did some googling and I'm fine with using a fromSQL to get data from my views.  I was able to add them to my context, albeit one of them I had to fake a primary key setup.  I don't have a particular view issue at this point in time - I can't really say what I was expecting.   Honestly I'm much more frustrated from not being able to inherit from a table model directly to a view model - but even that is more of an annoyance I worked around by creating a base model for them both to inherit from.

**Comment by nhustak** [comment]
@ajcvickers  I like where #9290 is headed.

**Comment by rnanjala** [comment]
@ajcvickers in the old entity framework with webforms, when I create the edmx, i am able to access the veiew directly like database objects but in the entity core 2.0, I couldn't do that directly. One of the workarounds I found is exxecuting fromsql(). is there a way, I could access the view directly like we do with tables?

**Comment by rogerhendriks** [comment]
@ajcvickers If #9290 supports mapping to read-only views and TVF's we can using it in a more professional way then with fromSQL.  I hope this will be in at first release? This and the groupby translation problems keeps us from starting full swing, so desparately waiting for 2.1 ;)
Maybe we need a seperate issue for Updatable views where you need to define a PK? Things can only become better as view support 6.x / edmx is not so good.

**Comment by Misiu** [comment]
@rogerhendriks I think updatable views could be done via stored procedure mapping https://github.com/aspnet/EntityFrameworkCore/issues/245.
I have a use case where I have table and view. View is basically `SELECT * FROM TABLE` but has some complex WHERE that filters rows in my view.
So ideally one could use view to read from database and stored procedures to do other operations (create,update,delete).

For example:
```c#
class BlogContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Blog>()
            .ToView("active_blogs")
			.WithStoredProcedures(s =>
				s.Update(u => u.HasName("modify_blog")
					.Parameter(b => b.BlogId, "blog_id")
					.Parameter(b => b.Name, "blog_name")
					.Parameter(b => b.Url, "blog_url"))
				 .Delete(d => d.HasName("delete_blog"))
				 .Insert(i => i.HasName("insert_blog")));
    }
}
```
similar to [EF6](https://msdn.microsoft.com/en-us/library/dn468673(v=vs.113).aspx)

**Comment by Misiu** [comment]
@ajcvickers any news about this? It is added to backlog, but I can't find any information about any progress.

**Comment by ajcvickers** [comment]
@Misiu When an issue is in the backlog it means that the EF team is intending to work on it at some point in the future, but is not actively working on it now--that's what the "backlog" is. For this issue specifically, there are several sub-issues, one of which is being worked on for 2.1--#9290. I believe the other sub-issues are also still in the backlog.

**Comment by eltiare** [comment]
I'm just here to add a "me too" to indicate that there is still interest in this feature.

**Comment by joswa888** [comment]
going 5 years.

**Comment by gsGabriel** [comment]
![image](https://user-images.githubusercontent.com/8701768/57253502-437ee000-7025-11e9-945b-bc86979a3c01.png)

i still waiting :D

**Comment by jhirschdrg** [comment]
As i see this will be one of the new feature in ef core 3
https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-3.0/features#reverse-engineering-of-database-views

**Comment by divega** [comment]
In 3.0 we have completed some of the most important aspects of this "aggregate issue" on database view support. Adding propose-close because I think its existence gets in the way of figuring out the relative priority of the remaining issues and even of coming up with other important things we could do. As far as i know all the remaining work that we know already has its own independent issues in the backlog. 

We could choose to make this a discussion issue, but I don't think those have proven very useful.

**Comment by mrlife** [comment]
After trying it out, https://github.com/dotnet/efcore/issues/9290 covers views well.

**Comment by DarioN1** [comment]
Please, have a look there:

https://github.com/DarioN1/SPToCore

I have created a kind of scaffolder for Stored Procedure, I'm using it in a new project that uses an existing database with more than 300 stored procedures.

**Comment by AraHaan** [comment]
8 years later, and still no direct view support in the scaffolder, nor the DbContext itself.


What a shame.

Like it's sad that one cant do:

```cs
// DiscordToken.cs
public partial record DiscordToken(
    [property: Key]
    int Id,
    string Token);

// BotDbContext.cs
public partial class BotDbContext : DbContext
{
    private readonly ILogger<BotDbContext> _logger;

    public BotDbContext(ILogger<BotDbContext> logger, DbContextOptions<BotDbContext> options)
        : base(options)
        => this._logger = logger;

    // table version.
    public virtual DbSet<DiscordToken> DiscordTokens { get; set; } = null!;
    // snip other tables.
    // view version that uses SELECT TOP 1.
    public virtual DbSet<DiscordToken> V_DiscordTokens { get; set; } = null!;
    // snip other views.
}
```

Instead what happens is that it throws ``SqlException: Invalid object name 'DiscordToken'.`` however I expect this code to "Just Work:tm:".

**Comment by roji** [comment]
@AraHaan can you provide a bit more info on how the above could "just work"? There's no view definition anywhere, how would EF Core know which view SQL it's supposed to create?

**Comment by AraHaan** [comment]
@roji the View itself already exists in the database (the database itself is managed by SSDT), I just want EFCore to simply be able to use the view. Currently the only option is to do
```cs
    public virtual DiscordToken GetDiscordToken()
    {
        this._logger.LogInformation("Executing [dbo].[V_DiscordTokens] view.");
        return RelationalQueryableExtensions.FromSqlRaw(
            this.DiscordTokens,
            "SELECT * FROM [dbo].[V_DiscordTokens]").AsNoTracking().First();
    }
```
inside of a ``BotDbContext.Views.cs`` file that is not touched by the scaffolding (which is why I make it partial).

And what I meant by could "just work" is that if the support was complete I could then drop the raw sql's in my code that is used to execute getting data from views.

**Comment by roji** [comment]
Are you looking for the [view mapping](https://docs.microsoft.com/en-us/ef/core/modeling/entity-types?tabs=data-annotations#view-mapping) features? Assuming your view already exists, you just need to tell EF which view to select from in the database, no raw SQL.

**Comment by ErikEJ** [comment]
@AraHaan EF Core just works with scaffolded views, they have HasNoKey().

Please create a new issue with full repro details.
